cpp/src/arrow/telemetry/logging.cc

```diff
@@ -0,0 +1,336 @@
+std::unique_ptr<otel::sdk::logs::LogRecordExporter> MakeExporter(
+    ExporterKind exporter_kind, std::ostream* ostream = nullptr) {
+  switch (exporter_kind) {
+    case ExporterKind::OSTREAM: {
+      return std::make_unique<otel::exporter::logs::OStreamLogRecordExporter>(*ostream);
+    } break;
+    case ExporterKind::OTLP_HTTP: {
+      namespace otlp = otel::exporter::otlp;
+      // TODO: Allow user configuration here?
+      otlp::OtlpHttpLogRecordExporterOptions options{};
+      return std::make_unique<otlp::OtlpHttpLogRecordExporter>(options);
+    } break;
+    case ExporterKind::OTLP_OSTREAM: {
+      return std::make_unique<OtlpOStreamLogRecordExporter>(ostream);
+    } break;
+    default:
+      break;
+  }
+  return nullptr;
+}
+
+std::unique_ptr<otel::sdk::logs::LogRecordExporter> MakeExporterFromEnv(
+    const LoggerProviderOptions& options) {
+  auto maybe_env_var = arrow::internal::GetEnvVar(kLoggingBackendEnvVar);
+  if (maybe_env_var.ok()) {
+    auto env_var = maybe_env_var.ValueOrDie();
+    auto* default_ostream =
+        options.default_export_stream ? options.default_export_stream : &std::cerr;
+    if (env_var == "ostream") {
+      // TODO: Currently disabled as the log records returned by otel's ostream exporter
+      // don't maintain copies of their attributes, leading to lifetime issues. If/when
+      // this is addressed, we can enable it. See:
+      // https://github.com/open-telemetry/opentelemetry-cpp/issues/2402
+#if 0
+      return MakeExporter(ExporterKind::OSTREAM, default_ostream);
+#else
+      ARROW_LOG(WARNING) << "Requested unimplemented backend " << kLoggingBackendEnvVar
+                         << "=" << env_var << ". Falling back to arrow_otlp_ostream";
+      return MakeExporter(ExporterKind::OTLP_OSTREAM, default_ostream);
+#endif
+    } else if (env_var == "otlp_http") {
+      return MakeExporter(ExporterKind::OTLP_HTTP);
+    } else if (env_var == "arrow_otlp_stdout") {
+      return MakeExporter(ExporterKind::OTLP_OSTREAM, &std::cout);
+    } else if (env_var == "arrow_otlp_stderr") {
+      return MakeExporter(ExporterKind::OTLP_OSTREAM, &std::cerr);
+    } else if (env_var == "arrow_otlp_ostream") {
+      return MakeExporter(ExporterKind::OTLP_OSTREAM, default_ostream);
+    } else if (!env_var.empty()) {
+      ARROW_LOG(WARNING) << "Requested unknown backend " << kLoggingBackendEnvVar << "="
+                         << env_var;
+    }
+  }
+  return nullptr;
+}
+
+std::unique_ptr<otel::sdk::logs::LogRecordProcessor> MakeLogRecordProcessor(
+    std::unique_ptr<otel::sdk::logs::LogRecordExporter> exporter) {
+  otel::sdk::logs::BatchLogRecordProcessorOptions options{};
+  return std::make_unique<otel::sdk::logs::BatchLogRecordProcessor>(std::move(exporter),
+                                                                    options);
+}
+
+otel::sdk::resource::Resource MakeResource(const ServiceAttributes& service_attributes) {
+  // TODO: We could also include process info...
```

Comment:
How could we get to a decision on whether to include these attributes? It seems like it could be useful information but, since users would typically be expected to correlate logs with traces, could this also be superfluous?

Reply:
> since users would typically be expected to correlate logs with traces, could this also be superfluous?

To clarify, is this because there would already be process info associated with the trace?

For the record, I don't think this is an ideal place to handle resource creation now that I'm looking at the big picture. This creates and sets attributes for a global OTel resource that _could_ be shared by both a `LoggerProvider` and `TracerProvider` - but it's currently hidden in the logging implementation exclusively.

Reply:
> To clarify, is this because there would already be process info associated with the trace?

Yep, that was my thought.

And your point about sharing makes sense to me. 👍

Reply:
Libraries shouldn't be setting this info. It should be configured once at the top level by the ultimate application. Arrow is just weird because in PyArrow you don't have a way to set it, but I wouldn't want to push this on C++ users.